File: App/Scripts/led.py

class Led:
    id: int = None
    r: int = 0
    g: int = 0
    b: int = 0
    a: float = .1

    def __init__(self, id: int = None, r: int = 0, g: int = 0, b: int = 0, a: float = .1):
        self.id = id
        self.r = r
        self.g = g
        self.b = b
        self.a = a
    # end_init

    def array(self) -> tuple[int, int, int, float]:
        return self.r, self.g, self.b, self.a
    # end_init

    def to_string(self) -> str:
        return '{id}: ({r}, {g}, {b}, {a})'.format(id=self.id, r=self.r, g=self.g, b=self.b, a=self.a)
    # end_init

    def serialize(self) -> str:
        return '{id},{r},{g},{b},{a}'.format(id=self.id, r=self.r, g=self.g, b=self.b, a=self.a)
    # end_init

    @staticmethod
    def deserialize(data: str):
        parts = data.split(',')
        if len(parts) != 5:
            return None

        try:
            return Led(
                id=None if parts[0] == 'None' else int(parts[0]),
                r=int(parts[1]),
                g=int(parts[2]),
                b=int(parts[3]),
                a=float(parts[4])
            )
        except (TypeError, ValueError):
            return None
        # end_try
    # end_init
# end_color

File: editor/api/led.php
<?php

class Led
{
    public ?int $id;
    public int $r = 0;
    public int $g = 0;
    public int $b = 0;
    public float $a = .1;

    public function __construct(?int $id = null, int $r = 0, int $g = 0, int $b = 0, float $a = .1)
    {
        $this->id = $id;
        $this->r = $r;
        $this->g = $g;
        $this->b = $b;
        $this->a = $a;
    }

    public function array(): array
    {
        return [
            $this->id,
            $this->r,
            $this->g,
            $this->b,
            $this->a,
        ];
    }

    public function toString(): string
    {
        return "{$this->id}: ({$this->r}, {$this->g}, {$this->b}, {$this->a})";
    }

    public function serialize(): string
    {
        return "{$this->id},{$this->r},{$this->g},{$this->b},{$this->a}";
    }

    public static function deserialize(string $data): Led | null
    {
        $parts = explode(',', $data);
        if (count($parts) != 5)
            return null;

        return new Led(
            id:(is_numeric($parts[0]) ? intval($parts[0]) : null),
            r:intval($parts[1]),
            g:intval($parts[2]),
            b:intval($parts[3]),
            a:intval($parts[4])
        );
    }
}
